Test with Windows on GitHub Actions

MaxDesiatov opened the pull request.

MaxDesiatov committed:
Test with Windows on GitHub Actions

broadwaylamb commented:
Awesome, thank you!

I can see that the C helpers aren’t compiling because of missing `pthread.h`. I think we could use `std::mutex` from C++ if pthread is not available.

MaxDesiatov commented:
Would you prefer `std::mutex` to be always used on all platforms, or to defer to `pthread` on platforms where the latter is available? I wonder if the former would be more convenient and reduce the amount if `#if` checks we have to do.

broadwaylamb commented:
Always using `std::mutex` would be convenient, but there are two reasons I haven’t already went with it:
- on Apple platforms, `os_unfair_lock` is much much faster than `pthread_mutex_lock` (and `std::mutex`, which uses `pthread_mutex_lock` on Darwin and Linux)
- pthread API allows to set the `PTHREAD_MUTEX_ERRORCHECK` attribute for mutexes, which enables checking for accidental recursive captures of non-recursive mutexes. `std::mutex` doesn’t allow that, and recursively capturing it is UB.

broadwaylamb committed:
Make COpenCombineHelpers buildable on Windows

broadwaylamb commented:
So, it now compiles, but we hit [this](https://github.com/apple/swift-tools-support-core/blob/492fe8dbf50c6f906444dbc227d3edb35ab0f4a7/Sources/TSCUtility/Triple.swift#L160).

broadwaylamb commented:
https://forums.swift.org/t/error-tscutility-triple-swift-fatal-error-could-not-determine-host-triple-malformed/47775/4

MaxDesiatov commented:
`import CoreFoundation` won't work on Windows since CoreFoundation is now hidden as a private implementation detail on non-Darwin platforms

broadwaylamb commented:
Well that's pretty stupid.

MaxDesiatov commented:
TWIMC, more details available here https://forums.swift.org/t/formalizing-the-unavailability-of-core-foundation/40216

broadwaylamb commented:
Thanks!

broadwaylamb committed:
I'm so tired of the Swift team breaking things on non-Darwin platforms

https://forums.swift.org/t/formalizing-the-unavailability-of-core-foundation/40216

broadwaylamb committed:
Support Windows threads in tests

broadwaylamb committed:
"Fix" Publishers.Breakpoint tests on Windows

broadwaylamb commented:
So, we now officially support Windows.

Congratulations and thank you!

broadwaylamb commented:
Let me know if it's safe to clean up the commit history — I don't want to force push to your branch.

MaxDesiatov commented:
Yeah, fine to clean up, no problem

broadwaylamb committed:
Support Windows in Package.swift

broadwaylamb committed:
Disable running tests on iOS 9.3

CircleCI have deprecated the image :(

broadwaylamb committed:
Mention Windows in README.md
